Snowplow/Internal/Utils/SPDataPersistence.m

```diff
@@ -0,0 +1,236 @@
+@property (nonatomic) NSString *userDefaultsKey;
+#endif
+
+@property (nonatomic) NSURL *directoryUrl;
+@property (nonatomic) NSURL *fileUrl;
+
+@end
+
+@implementation SPDataPersistence
+
+static NSMutableDictionary<NSString *, SPDataPersistence *> *instances = nil;
+
+#if TARGET_OS_TV || TARGET_OS_WATCH
+NSString *const kSPSessionDictionaryPrefix = @"SPSessionDictionary";
+#endif
+
+NSString *const kFilename = @"namespace";
+NSString *const kFilenameExt = @"dict";
+NSString *const kSessionFilenameV1 = @"session.dict";
+NSString *const kSessionFilenamePrefixV2_2 = @"session";
+
+NSString *sessionKey = @"session";
+
++ (SPDataPersistence *)dataPersistenceForNamespace:(NSString *)namespace {
+    NSString *escapedNamespace = [SPDataPersistence stringFromNamespace:namespace];
+    if ([escapedNamespace length] <= 0) return nil;
+    @synchronized (SPDataPersistence.class) {
+        SPDataPersistence *instance = nil;
+        if (instances) {
+            instance = [instances objectForKey:escapedNamespace];
+            if (instance) {
+                return instance;
+            }
+        } else {
+            instances = [NSMutableDictionary new];
+        }
+        instance = [[SPDataPersistence alloc] initWithNamespace:escapedNamespace];
+        [instances setValue:instance forKey:escapedNamespace];
+        return instance;
+    }
+}
+
++ (BOOL)removeDataPersistenceWithNamespace:(NSString *)namespace {
+    SPDataPersistence *instance = [SPDataPersistence dataPersistenceForNamespace:namespace];
+    if (!instance) return NO;
+    @synchronized (SPDataPersistence.class) {
+        [instances removeObjectForKey:instance.escapedNamespace];
+    }
+    [instance removeAll];
+    return YES;
+}
+
++ (NSString *)stringFromNamespace:(NSString *)namespace {
+    if (!namespace) return nil;
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"[^a-zA-Z0-9_]+" options:0 error:nil];
+    return [regex stringByReplacingMatchesInString:namespace options:0 range:NSMakeRange(0, namespace.length) withTemplate:@"-"];
+}
+
+// MARK: - Property accessor methods
+
+- (NSDictionary<NSString *, NSDictionary<NSString *, NSObject *> *> *)data {
+    @synchronized (self) {
+#if TARGET_OS_TV || TARGET_OS_WATCH
+        return [[NSUserDefaults standardUserDefaults] dictionaryForKey:self.userDefaultsKey];
+#else
+        NSMutableDictionary<NSString *, NSDictionary<NSString *, NSObject *> *> *result =
+        [NSMutableDictionary dictionaryWithContentsOfURL:self.fileUrl];
+        
+        if (!result) {
+            // Initialise
+            result = [NSMutableDictionary new];
+            // Migrate legacy session data
+            NSDictionary *sessionDict = [self sessionDictionaryFromLegacyTrackerV2_2]
+                ?: [self sessionDictionaryFromLegacyTrackerV1]
+                ?: [NSDictionary new];
+            [result setObject:sessionDict forKey:sessionKey];
+            [self storeDictionary:result];
+        }
```

Comment:
I don't think it needs doing for this release but if theres ever another migration layer then I think this might need extracting out of here into something thats easier to extend.

Reply:
I agree. The problem of the migration has emerged often recently. This DataPersistence component may be the place where to move all the filesystem management needed in the tracker and related migration (with a clearer strategy).